lib/services/extensibility-service.ts

```diff
@@ -0,0 +1,135 @@
+
+		const localPath = path.resolve(extensionName);
+		const packageName = this.$fs.exists(localPath) ? localPath : extensionName;
+
+		const realName = (await this.$npm.install(packageName, this.pathToExtensions, npmOpts))[0];
+		this.$logger.trace(`Finished installation of extension '${extensionName}'. Trying to load it now.`);
+
+		// In case the extension is already installed, the $npm.install method will not return the name of the package.
+		// Fallback to the original value.
+		// NOTE: This will not be required once $npm.install starts working correctly.
+		return await this.loadExtension(realName || extensionName);
+	}
+
+	public async uninstallExtension(extensionName: string): Promise<void> {
+		this.$logger.trace(`Start uninstallation of extension '${extensionName}'.`);
+
+		await this.assertExtensionsSetup();
+
+		await this.$npm.uninstall(extensionName, { save: true }, this.pathToExtensions);
+
+		this.$logger.trace(`Finished uninstallation of extension '${extensionName}'.`);
+	}
+
+	public loadExtensions(): Promise<any>[] {
+		this.$logger.trace("Loading extensions.");
+
+		let dependencies: IStringDictionary = null;
+
+		try {
+			dependencies = this.getInstalledExtensions();
+		} catch (err) {
+			this.$logger.trace(`Error while getting installed dependencies: ${err.message}. No extensions will be loaded.`);
+		}
+
+		return _.keys(dependencies)
+			.map(name => this.loadExtension(name));
+	}
+
+	public getInstalledExtensions(): IStringDictionary {
+		if (this.$fs.exists(this.pathToPackageJson)) {
+			return this.$fs.readJson(this.pathToPackageJson).dependencies;
+		}
+
+		return null;
+	}
+
+	private async loadExtension(extensionName: string): Promise<IExtensionData> {
+		try {
+			await this.assertExtensionIsInstalled(extensionName);
+
+			const pathToExtension = path.join(this.pathToExtensions, constants.NODE_MODULES_FOLDER_NAME, extensionName);
+			this.$requireService.require(pathToExtension);
+			return { extensionName };
+		} catch (error) {
+			this.$logger.warn(`Error while loading ${extensionName} is: ${error.message}`);
+			const err = <IExtensionLoadingError>new Error(`Unable to load extension ${extensionName}. You will not be able to use the functionality that it adds.`);
+			err.extensionName = extensionName;
+			throw err;
+		}
+	}
+
+	private async assertExtensionIsInstalled(extensionName: string): Promise<void> {
+		this.$logger.trace(`Asserting extension ${extensionName} is installed.`);
+		const installedExtensions = this.$fs.readDirectory(path.join(this.pathToExtensions, constants.NODE_MODULES_FOLDER_NAME));
+
+		if (installedExtensions.indexOf(extensionName) === -1) {
+			this.$logger.trace(`Extension ${extensionName} is not installed, starting installation.`);
+			await this.installExtension(extensionName);
+		}
+
+		this.$logger.trace(`Extension ${extensionName} is installed.`);
+	}
+
+	@cache()
+	private async assertExtensionsSetup(): Promise<void> {
+		this.assertExtensionsDirExists();
```

Comment:
From a semantic point of view I think it'd be more natural if `this.assertPackageJsonExists` calls `this.assertExtensionsDirExists` and you just call `this.assertPackageJsonExists` here.
That's my 2¢ of course hence definitely not a merge-stopper